hypothesis-python/src/hypothesis/internal/conjecture/shrinking/dfas.py

```diff
@@ -0,0 +1,336 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Most of this work is copyright (C) 2013-2020 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+import hashlib
+import math
+from itertools import islice
+
+from hypothesis import HealthCheck, settings
+from hypothesis.errors import HypothesisException
+from hypothesis.internal.conjecture.data import ConjectureResult, Status
+from hypothesis.internal.conjecture.dfa.lstar import LStar
+from hypothesis.internal.conjecture.shrinking.learned_dfas import (
+    SHRINKING_DFAS,
+    __file__ as learned_dfa_file,
+)
+
+"""
+This is a module for learning new DFAs that help normalize test
+functions. That is, given a test function that sometimes shrinks
+to one thing and sometimes another, this module is designed to
+help learn new DFA-based shrink passes that will cause it to
+always shrink to the same thing.
+"""
+
+
+class FailedToNormalise(HypothesisException):
+    pass
+
+
+def update_learned_dfas():
+    """Write any modifications to the SHRINKING_DFAS dictionary
+    back to the learned DFAs file."""
+
+    with open(learned_dfa_file) as i:
+        source = i.read()
+
+    lines = source.splitlines()
+
+    i = lines.index("# AUTOGENERATED BEGINS")
+
+    del lines[i + 1 :]
+
+    lines.append("")
+    lines.append("# fmt: off")
+    lines.append("")
+
+    for k, v in sorted(SHRINKING_DFAS.items()):
+        lines.append("SHRINKING_DFAS[%r] = %r  # noqa: E501" % (k, v))
+
+    lines.append("")
+    lines.append("# fmt: on")
+
+    new_source = "\n".join(lines) + "\n"
+
+    if new_source != source:
+        with open(learned_dfa_file, "w") as o:
+            o.write(new_source)
+
+
+def learn_a_new_dfa(runner, u, v, predicate):
+    """Given two buffers ``u`` and ``v```, learn a DFA that will
+    allow the shrinker to normalise them better. ``u`` and ``v``
+    should not currently shrink to the test case when calling
```

Comment:
Is this supposed to say `the [same] test case`?

Reply:
Yes.